unstructured/ingest/v2/processes/connectors/salesforce.py

```diff
@@ -0,0 +1,293 @@
+            return self.private_key, Path
+
+        raise ValueError("private_key does not contain PEM private key or path")
+
+
+@dataclass
+class SalesforceConnectionConfig(ConnectionConfig):
+    username: str
+    access_config: SalesforceAccessConfig = enhanced_field(sensitive=True)
+
+    @requires_dependencies(["simple_salesforce"], extras="salesforce")
+    def get_client(self):
+        from simple_salesforce import Salesforce
+
+        pkey_value, pkey_type = self.access_config.get_private_key_value_and_type()
+
+        return Salesforce(
+            username=self.username,
+            consumer_key=self.access_config.consumer_key,
+            privatekey_file=pkey_value if pkey_type is Path else None,
+            privatekey=pkey_value if pkey_type is str else None,
+            version=SALESFORCE_API_VERSION,
+        )
+
+
+@dataclass
+class SalesforceIndexerConfig(IndexerConfig):
+    categories: t.List[str]
+
+
+@dataclass
+class SalesforceIndexer(Indexer):
+    connection_config: SalesforceConnectionConfig
+    index_config: SalesforceIndexerConfig
+
+    @requires_dependencies(["simple_salesforce"], extras="salesforce")
+    def list_files(self) -> t.List[FileData]:
+        """Get Salesforce Ids for the records.
+        Send them to next phase where each doc gets downloaded into the
+        appropriate format for partitioning.
+        """
+        from simple_salesforce.exceptions import SalesforceMalformedRequest
+
+        client = self.connection_config.get_client()
+
+        files_list = []
+        for record_type in self.index_config.categories:
+            if record_type not in ACCEPTED_CATEGORIES:
+                raise ValueError(f"{record_type} not currently an accepted Salesforce category")
+
+            try:
+                # Get ids from Salesforce
+                records = client.query_all_iter(
+                    f"select Id, SystemModstamp, CreatedDate, LastModifiedDate from {record_type}",
+                )
+                for record in records:
+                    files_list.append(
+                        FileData(
+                            connector_type=CONNECTOR_TYPE,
+                            identifier=record["Id"],
+                            source_identifiers=SourceIdentifiers(
+                                filename=record["Id"],
+                                fullpath=f"{record['attributes']['type']}/{record['Id']}",
+                            ),
+                            metadata=DataSourceMetadata(
+                                url=record["attributes"]["url"],
+                                version=record["SystemModstamp"],
+                                date_created=record["CreatedDate"],
+                                date_modified=record["LastModifiedDate"],
+                                record_locator={"id": record["Id"]},
+                            ),
+                            additional_metadata={"record_type": record["attributes"]["type"]},
```

Comment:
I've been adding whatever metadata comes back from the provider that isn't already being set in the metadata here, as long as it's json serializable. You never know what might need that data downstream and it won't be added to every element, so it won't bloat the partitioned output.

Reply:
we choose what fields to query for this. I'd rather keep the query to a minimum of what we need. salesforce doesn't have a `select * from` for this query.